### ISSUE TITLE
Typeorm generating migrations even though there is not any change in  models

### ISSUE THREAD

**Issue by chandrashekhar07** [typeorm/typeorm#7955]
## Issue Description
i created a passenger.entity.ts file added few fields over it.
Then i generated a migration named first and i run migration run command successfully and then i can feel changes to my PostgreSQL database,

after few minutes, again i generated a migration named second,

here i am surprised to see that the second migration even generate the same SQL query as of first migrations. 

### Expected Behavior

i believe when i run command to generated second migration, it should show "typeORM No changes in database schema were found - cannot generate a migration. To create a new empty migration use “typeorm migration:create” command "

### Actual Behavior

**when i run first migration, following code is generated:**

import {MigrationInterface, QueryRunner} from "typeorm";

export class first1627232991461 implements MigrationInterface {
    name = 'first1627232991461'

    public async up(queryRunner: QueryRunner): Promise<void> {
        await queryRunner.query(`ALTER TABLE "passengers" ALTER COLUMN "is_deleted" SET DEFAULT 'false'`);
    }

    public async down(queryRunner: QueryRunner): Promise<void> {
        await queryRunner.query(`ALTER TABLE "passengers" ALTER COLUMN "is_deleted" SET DEFAULT false`);
    }

}

**when i run second migration command, following code is genetated:**

import {MigrationInterface, QueryRunner} from "typeorm";

export class second1627234626276 implements MigrationInterface {
    name = 'second1627234626276'

    public async up(queryRunner: QueryRunner): Promise<void> {
        await queryRunner.query(`ALTER TABLE "passengers" ALTER COLUMN "is_deleted" SET DEFAULT 'false'`);
    }

    public async down(queryRunner: QueryRunner): Promise<void> {
        await queryRunner.query(`ALTER TABLE "passengers" ALTER COLUMN "is_deleted" SET DEFAULT false`);
    }

}

Surprisingly in both case there is same sql query

### Steps to Reproduce

name of my column is inside .entity.ts file is:

    @Column({ name: 'is_deleted', type: 'boolean', default: 'false' })
    public isDeleted: boolean;

### My Environment

| Dependency          | Version  |
| ---                 | ---      |
| Operating System    |   Ubuntu 20.04       |
| Node.js version     | 14.16.1  | 
| Typescript version  | 4..2.4  | 
| TypeORM version     | 0.2.32 | 

### Additional Context
i believe following image clearly show this issue

![image](https://user-images.githubusercontent.com/35414590/126908196-58072023-93cf-4b9d-af1c-21764fe1d58d.png)

### Relevant Database Driver(s)

| DB Type              | Reproducible  |
| ---                  | ---           |
| `postgres`           | yes            |

### Are you willing to resolve this issue by submitting a Pull Request?

 -  ✖️ Yes, I have the time, and I know how to start.
 - ✅ Yes, I have the time, but I don't know how to start. I would need guidance.
 - ✖️ No, I don’t have the time, but I can support (using donations) development.
 - ✖️ No, I don’t have the time and I’m okay to wait for the community / maintainers to resolve this issue.

**Comment by imnotjames** [comment]
Best guess without digging into it:

The `false` is normalized by the postgres database server to be `FALSE` which `FALSE !== false` so it generates a new one.

**Comment by chandrashekhar07** [comment]
@imnotjames , i tried changing default value to FALSE, but still same sql query is regenerated,

**Comment by omen-x** [comment]
Got the same problem. The "duplicate" query in my case is:
```SQL
await queryRunner.query(`
            ALTER TABLE "foo"
            ALTER COLUMN "bar"
            SET DEFAULT to_jsonb('{}'::
`);

```
```javascript
@Column({ type: "jsonb", name: "bar", default: () => "to_jsonb('{}'::text)" })
public bar: string;

```

Node - 12.18.3
TS - 3.7
Typeorm - 0.2.32
Postgres

**Comment by imnotjames** [comment]
> Got the same problem. The "duplicate" query in my case is:
> 
> ```sql
> await queryRunner.query(`
>             ALTER TABLE "foo"
>             ALTER COLUMN "bar"
>             SET DEFAULT to_jsonb('{}'::
> `);
> ```
> 
> ```js
> @Column({ type: "jsonb", name: "bar", default: () => "to_jsonb('{}'::text)" })
> public bar: string;
> ```
> 
> Node - 12.18.3
> TS - 3.7
> Typeorm - 0.2.32
> Postgres

This isn't the same problem and you aren't using the latest release.

**Comment by lflfm** [comment]
it's not clear on your post, looks like you didn't....
; did you actually run the migration or just the generate command?

the generate command just generates code by looking at the db. If you dont use the `run` command, the db is unchanged and the next generate will find the db in the same state 😊

**Comment by chandrashekhar07** [comment]
@lflfm i have run the migration command (after generating it),
and even after running the migration command, 
Migration:generate command is generating migration files which contains same SQL query as of first migration.

**Comment by lflfm** [comment]
_which_ migration command? did you run `migration:run` ?
correct order is:
`typeorm-cli migration:generate -n mynewChanges`
`typeorm-cli migration:run`
`typeorm-cli migration:generate -n someMoreChanges`

**Comment by chandrashekhar07** [comment]
@lflfm  yeah  i know the correct order and i have followed the same order (you can check on photo attached above, i have merged generate and run command with && )
here, 
``` typeorm-cli migration:generate -n someMoreChanges ``` is generating same sql files as of ``` typeorm-cli migration:generate -n mynewChanges ```

**Comment by lflfm** [comment]
Sorry, I hadn't read the command completely.
The only thing I notice now then is that the migrate up is trying to set default to a `string` that says `'false'` while the migrate down sets the default to a `boolean` `false`. Maybe that's the issue?
Have you tried to set the default as a boolean in your model?
instead of:
`@Column({ name: 'is_deleted', type: 'boolean', default: 'false' })`
try:
`@Column({ name: 'is_deleted', type: 'boolean', default: false })`

**Comment by chandrashekhar07** [comment]
@lflfm Thank u man, setting ```default:false``` solves the issue

**Comment by forrestwilkins** [comment]
I'm getting a similar issue to this. I get the following SQL generated, even though no changes have been made to the `GroupMemberRequest` entity in ages:
```
await queryRunner.query(
  `CREATE SEQUENCE IF NOT EXISTS "group_member_request_id_seq" OWNED BY "group_member_request"."id"`
);
await queryRunner.query(
  `ALTER TABLE "group_member_request" ALTER COLUMN "id" SET DEFAULT nextval('"group_member_request_id_seq"')`
);
await queryRunner.query(
  `ALTER TABLE "group_member_request" ALTER COLUMN "id" DROP DEFAULT`
);
```
I've cleared out my `dist` folder and also tried recreating my local dev database with no luck.

**Comment by forrestwilkins** [comment]
Has anyone tried the solution found [here](https://stackoverflow.com/questions/73807045/typeorm-0-3-migration-generates-changes-even-without-making-a-single-change-to-t) by chance?

**Comment by CodySwannGT** [comment]
Having the same issue

**Comment by sjlehn** [comment]
> Has anyone tried the solution found [here](https://stackoverflow.com/questions/73807045/typeorm-0-3-migration-generates-changes-even-without-making-a-single-change-to-t) by chance?

I just did. It didn't help. I still have a pile of bogus migrations that re-create a handful of indexes and foreign keys, as well as some regular fields. There's no difference in what's in `up` vs what's in `down`. When I dig into the debugger, it looks like it thinks the FKs should be cascading when they're not (in `node_modules/typeorm/schema-builder/RdbmsSchemaBuilder.js::dropOldForeignKeys`), though the generated update shows NO ACTION, which matches the existing table.

Some of the fields being re-created are shared by 40 some tables (they're history fields - createdAt, modifiedAt, etc), yet only 3 tables are getting fake migrations generated. I can't figure out what makes those particular tables unique.

**Comment by zedjarvis** [comment]
I realized that the `migration:generate` will re-create the same migration only if you have not executed `migrations:run`. After running migration then try to re-generate without changes you will get the following message `No changes in database schema were found - cannot generate a migration. To create a new empty migration use "typeorm migration:create" command`

**Comment by sjlehn** [comment]
Our team found the source of the additional migrations in our project; it was a configuration issue. We were defining ManyToMany relationships multiple times for a single relation, once with `@ManyToMany` from entity to entity, and again with `@JoinTable` on both sides that pointed to the junction table. This resulted in TypeORM creating two entities for every relation, and any discrepancy between the two caused the additional migrations. If you're defining manual junction tables via `@JoinTable` (b/c you need additional data there), you can't use `@ManyToMany`, and vice-versa.

**Comment by ironclock** [comment]
Years later, this is still an issue smh
